ci/jenkins/scale-agent.sh

```diff
@@ -0,0 +1,285 @@
+}
+
+WORKDIR="/var/lib/jenkins"
+RUN_SETUP_ONLY=false
+RUN_CLEANUP_ONLY=false
+JENKINS_URL="https://jenkins.antrea.io"
+LABEL="kind"
+ZONE="us-west1-a"
+MACHINE_TYPE="e2-standard-4"
+IMAGE_FAMILY="ubuntu-2204-lts"
+IMAGE_PROJECT="ubuntu-os-cloud"
+BOOT_DISK_SIZE="200GB"
+AGENT_NAME_PATTERN="jenkins-agent"
+NEW_AGENT_NAME="$AGENT_NAME_PATTERN-$(date +%s)"
+
+_usage="Usage: $0 [--workdir <HomePath>] [--setup-only] [--cleanup-only] [--kind]
+
+Scale a jenkins agent to run CI tests.
+
+        --workdir                Home path for Go, vSphere information and antrea_logs during cluster setup. Default is $WORKDIR.
+        --setup-only             Only perform setting up the cluster and run test.
+        --cleanup-only           Only perform cleaning up the cluster.
+        --jenkins-user           Jenkins user name.
+        --jenkins-token          Jenkins API token.
+        --label                  Label for the jenkins agent.
+        --kind                   Setup kind testbed."
+
+function print_usage {
+    echoerr "$_usage"
+}
+
+function print_help {
+    echoerr "Try '$0 --help' for more information."
+}
+
+while [[ $# -gt 0 ]]
+do
+key="$1"
+
+case $key in
+    --workdir)
+    WORKDIR="$2"
+    shift 2
+    ;;
+    --setup-only)
+    RUN_SETUP_ONLY=true
+    shift
+    ;;
+    --cleanup-only)
+    RUN_CLEANUP_ONLY=true
+    shift
+    ;;
+    --kind)
+    KIND=true
+    shift
+    ;;
+    --jenkins-user)
+    JENKINS_USER="$2"
+    shift 2
+    ;;
+    --jenkins-token)
+    JENKINS_TOKEN="$2"
+    shift 2 
+    ;;
+    --label)
+    LABEL="$2"
+    shift 2
+    ;;
+    -h|--help)
+    print_usage
+    exit 0
+    ;;
+    *)    # unknown option
+    echoerr "Unknown option $1"
+    exit 1
+    ;;
+esac
+done
+
+function check_jobs_in_queue {
+    local LABEL=$1
+    local SUCCESS=0
+
+    for i in {1..2}; do
+        QUEUE_JOBS=$(curl -k -s -u "$JENKINS_USER:$JENKINS_API_TOKEN" "$JENKINS_URL/queue/api/json")
+
+        JOB_FOUND=$(echo "$QUEUE_JOBS" | jq -r '.items[] | select(.task.labels != null) | .task.labels[]' | grep -w "$LABEL")
+
+        if [ -n "$JOB_FOUND" ]; then
+            echo "Job with label $LABEL found in queue. Retrying in 10 seconds..."
+            SUCCESS=1
+            sleep 10
+        else
+            echo "No job with label $LABEL found in queue. Exit."
+            SUCCESS=0
+            break
+        fi
+    done
+
+    echo $SUCCESS
+}
+
+function add_agent {
+    echo "Checking if there are jobs in the Jenkins queue with label: $LABEL"
+    JOB_EXISTS=$(check_jobs_in_queue "$LABEL")
+
+    if [ "$JOB_EXISTS" -ne 0 ]; then
+        echo "Jobs with label $LABEL not found in the queue. Exit."
+        return 1
+    fi
+
+    echo "Jobs with label $LABEL exist in the queue. Proceeding with agent creation..."
+    sudo gcloud compute instances create "$NEW_AGENT_NAME" --zone="$ZONE" --machine-type="$MACHINE_TYPE" --image-family="$IMAGE_FAMILY" --image-project="$IMAGE_PROJECT" --boot-disk-size="$BOOT_DISK_SIZE"
```

Comment:
It's necessary to restrict the maximum number of agent VMs. Not sure if we can also configure a customized quota in GCP for the CI user, too. This is to avoid creating too many VMs by accident.

Are openjdk, golang, make, gcc, docker and kind installed in advance in the VM? I didn't find how they are installed in this script.

Reply:
As we discussed, I have created quota in GCP for VM restriction, also updated jdk installation for join the command.

Reply:
Quota is just a fail-safe. I think the script should try to limit the overall CI VM count. This is because later if developers want to deploy some VMs on GCP for testing purpose, the quota will be raised, and it will be more than the CI VMs we want to deploy. Suppose that the developer deleted the dev VMs after use, the quota will be available for CI VMs. This may cause CI to deploy too many CI VMs.

Reply:
I see, it's reasonable to add this parameter, added `max-agents` for this requirement.